File: session-notes/2015-02-20/topics-in-cs/rpn/rpn.js
// var RPN = {
//   "OPERATORS": {

//   },
//   "Expression": function() {
//     return {
//       evaluate: function(){}
//     }
//   }
// };

// (new RPN.Expression("3 4 -")).evaluate();


var OPERATORS = {
  "+": "+",
  "-": "-",
  "*": "*",
  "/": "/",
  "%": "%"
};

// use regex? //slower than hash table

// empty string is false
// typeof "" is string
var rpn = function rpn(string, stack) {
  console.log(string, typeof string);
  console.log("stack:", stack);
  var args = typeof string === "string" ? string.split(" ") : string;
  var count = args.length;
  if (!stack) {
    var stack = [];
    console.log("no stack. created empty stack")
  }

  if ( (typeof string === 'object' || !string.length) && !stack.length) {
    console.log("if branch");
    return null;
  } else if ((typeof string === 'object' || !string.length) && stack.length === 1) {
    console.log("first else if branch");
    console.log("stack:", stack);
    console.log("stack[0]", stack[0]);
    return stack[0]; //undefined why?!!!
  } else {
    console.log("second else if branch");

    var arg = args.shift();
    if (OPERATORS[arg]) {
      console.log("operator")
      var right = stack.pop();
      var left = stack.pop();
      var calculatedValue = eval(left.concat(arg).concat(right));
      console.log("left:", left, "operator:", arg, "right:", right, "calculatedValue:", calculatedValue);
      stack.push((calculatedValue).toString());
      rpn(args, stack);
    } else {
      // assuming it is a typeof "number"
      stack.push(arg);
      args = args.join(" ");
      rpn(args, stack);
    }
  }
};

File: session-notes/2015-02-20/topics-in-cs/rpn/rpn.rb
module RPN
  require 'set'

  OPERATORS = Set.new(%w[+ - / * ** %])

  class Expression

    def initialize(string)
      @tokens = string.split(" ")
      @stack = []
    end

    def evaluate
      count = @tokens.count
      # puts "count: #{count}"
      # puts "stack: #{@stack}"
      if count == 0 && @stack.count == 0
        nil
      elsif count == 0 && @stack.count == 1
        @stack[0]
      else
        arg = @tokens.shift
        if OPERATORS.include?(arg)
          # puts "operator #{arg}"
          right = @stack.pop
          left = @stack.pop
          @stack << eval(left.concat(arg).concat(right)).to_s
          self.evaluate
        else #assumes it is a valid number
          @stack << arg
          self.evaluate
        end

      end

    end

  end

end
